### ISSUE TITLE
Why are just the embedding used for inference and not the finetuned model itself?

### ISSUE THREAD

**Issue by datistiquo** [UKPLab/sentence-transformers#540]
Hello,

thank you for this repo and you nice paper. :)

I have an understanding question about the paper and the usage of sentence embeddings. Sort: Why are you using the embeddings and then the cosine distance instead of the finetuned model for predictions? As I understand you finetune the siamese model on the data, but for making predictions you do not use this model, instead just extracting the emebddings and calculating plain cosine. 

Is there any reason for this? Or maybe this method is better to handle overfitting/generalizability?

I try to get a siamese model too for my own data, but I am far away to reach the same performance of the plain bert with a simple classifier on top if feeding both sentences as input. So, decoupling the sentences and use them as seperated input to BERT somehow breaks drastically the performance. If I use the plain BERT feeding both sentences as input with 2-way softmax classifier on top I get almost perfect results and generalizability is also very good. But decoupling in a siamese manner (and a 2-softmax layer on top) is not even converging.
So finetuning like the huggingface TFBERTForSequenceClassifciation (BERT model with a simple softmax layer like above) with both inputs is perfect, but finetuning same model in siamese manner (decoupling) is not even converging and you nneed a larger neural Network before the classifier softmax which yields also good performance but overfitting is an issue.

I assume this is because of the kind of pretraining of BERT? Since BERT is trained also with 2 sentences as input, it is almost better when use with 2 sentences.

**Comment by nreimers** [comment]
Have a look here in the intro for the motivation for this:
https://arxiv.org/abs/1908.10084

Assume you have 100k questions in your community board and you want to find all duplicate questions to merge them. With BERT, you would need to score about 100k x 100k question pairs, which will take forever.

With SBERT, you compute the embedding for 100k sentences, then compute the cosine similarity and you mark all pairs with e.g. a score >0.9 as duplicate. This can be done in less 100 seconds.


If you have just fixed pairs, then traditional BERT is better. If your task is clustering, information retrieval, or paraphrase mining, then BERT is not suitable as the run time is way too slow.

**Comment by datistiquo** [comment]
So, also with BERT and single input sentence the purpose was time? Because for my case the reason is that I have some fixed sentences where I can encode them beforehand and then using just bert for a single query which is fast. But using BERT with both as input I would have to encode all possible sentences which is slow of course. For me that is the only reason to use BER with a single input, but this drops performance drastically as BERT might be doing better with dual input?

So you just considered time to use the finetuned embeddings and not the model itself with a classifier on top? 

Any hints if overfitting of a finetuned model is propagate to the embeddings itself? If the finetuned model is bad on some cases are the embeddings itself bad too?

**Comment by nreimers** [comment]
If you have a fixed (small) number of pairs, bi-encoders like SBERT are not needed and cross-encoder based on BERT achieve better performances:
https://www.sbert.net/docs/usage/cross-encoder.html

**Comment by datistiquo** [comment]
Thank you. I already came to the same conclusion by my use case. As I have limits such that that cross encoder is also not practical in my case I need to have such a siamese bi model or using the embedings from the finetuned model. 

And, I thought it should play no role if you use the finetuned model at whole or just the emebddings for  even a large set of questions, because you "just" have a simple softmax layer on top?! I think feed the input to this layer or calculating the cosine should be the same. The most time consuming part is feed it to the bert model and this you do in both cases...
